File: backend/app/services/balance_service.py
from sqlalchemy.orm import Session
from app.models.group_member import GroupMember
from app.models.group_balance import GroupBalance
from app.models.split import Split
from app.models.settlement import Settlement
from app.models.expense_item import ExpenseItem
from app.models.expense import Expense
from app.models.user import User
from decimal import Decimal
from datetime import datetime
import uuid


def compute_group_balances(db: Session, group_id: uuid.UUID) -> list[dict]:
    """
    Compute net balance for each user in a group.
    Returns details: paid, share, net balance.
    """
    members = db.query(GroupMember, User).join(
        User, GroupMember.user_id == User.id
    ).filter(GroupMember.group_id == group_id).all()
    
    balances = {}
    for member, user in members:
        balances[member.user_id] = {
            "user_id": member.user_id,
            "user_name": user.name,
            "total_paid": Decimal("0"),
            "total_share": Decimal("0"),
            "total_settled": Decimal("0"),  # Positive = Received, Negative = Paid
            "net_balance": Decimal("0"),
        }
    
    expenses = db.query(Expense).filter(Expense.group_id == group_id).all()
    
    for expense in expenses:
        # Skip expenses that are still processing (total_amount is NULL)
        if expense.total_amount is None:
            continue
            
        # Calculate total based on actual splits to ensure system balances (Net = 0)
        # Verify payer exists in group (handle case if payer left but splits remain)
        payer_id = expense.created_by
        
        # We need to sum up all splits for this expense
        expense_total_shares = Decimal("0")
        
        items = db.query(ExpenseItem).filter(ExpenseItem.expense_id == expense.id).all()
        
        for item in items:
            splits = db.query(Split).filter(Split.expense_item_id == item.id).all()
            
            for split in splits:
                if split.user_id in balances:
                    balances[split.user_id]["total_share"] += split.amount
                    balances[split.user_id]["net_balance"] -= split.amount
                    expense_total_shares += split.amount

        # Credit the payer with the EXACT amount that was split among members
        # This ensures Sum(Net Balances) == 0, preventing "phantom debt"
        if payer_id and payer_id in balances:
            balances[payer_id]["total_paid"] += expense_total_shares
            balances[payer_id]["net_balance"] += expense_total_shares
    
    settlements = db.query(Settlement).filter(Settlement.group_id == group_id).all()
    
    for settlement in settlements:
        if settlement.paid_by in balances:
            balances[settlement.paid_by]["net_balance"] += settlement.amount
            balances[settlement.paid_by]["total_settled"] -= settlement.amount
        
        if settlement.paid_to in balances:
            balances[settlement.paid_to]["net_balance"] -= settlement.amount
            balances[settlement.paid_to]["total_settled"] += settlement.amount
    
    return list(balances.values())


def update_group_balances(db: Session, group_id: uuid.UUID):
    """
    Update the group_balances cache table. 
    """
    computed_balances = compute_group_balances(db, group_id)
    
    for balance_data in computed_balances:
        existing_balance = db.query(GroupBalance).filter(
            GroupBalance.group_id == group_id,
            GroupBalance.user_id == balance_data["user_id"]
        ).first()
        
        if existing_balance:
            existing_balance.net_balance = balance_data["net_balance"]
            existing_balance.updated_at = datetime.utcnow()
        else:
            new_balance = GroupBalance(
                group_id=group_id,
                user_id=balance_data["user_id"],
                net_balance=balance_data["net_balance"],
                updated_at=datetime.utcnow()
            )
            db.add(new_balance)
    
    db.commit()


def minimize_debts(balances: list[dict]) -> list[dict]:
    """
    Simplify debts using a greedy algorithm.
    Input balances have 'net_balance' and 'user_name'.
    """
    debtors = []
    creditors = []
    
    for b in balances:
        net = b["net_balance"]
        if net < 0:
            debtors.append({"user_id": b["user_id"], "user_name": b["user_name"], "amount": -net})
        elif net > 0:
            creditors.append({"user_id": b["user_id"], "user_name": b["user_name"], "amount": net})
            
    # Sort by amount descending to greedy match
    debtors.sort(key=lambda x: x["amount"], reverse=True)
    creditors.sort(key=lambda x: x["amount"], reverse=True)
    
    debts = []
    i = 0 # debtor index
    j = 0 # creditor index
    
    while i < len(debtors) and j < len(creditors):
        debtor = debtors[i]
        creditor = creditors[j]
        
        # Amount to settle is min of what debtor owes and creditor is owed
        amount = min(debtor["amount"], creditor["amount"])
        
        if amount > 0:
            debts.append({
                "from_user": debtor["user_name"],
                "to_user": creditor["user_name"],
                "amount": round(amount, 2)
            })
            
        # Update remaining amounts
        debtor["amount"] -= amount
        creditor["amount"] -= amount
        
        # Move indices if settled
        if debtor["amount"] < 0.01:
            i += 1
        if creditor["amount"] < 0.01:
            j += 1
            
    return debts


def calculate_group_debts(db: Session, group_id: uuid.UUID) -> dict:
    """
    Return comprehensive balance view: user stats AND simplified debts.
    """
    balances_list = compute_group_balances(db, group_id)
    debts_list = minimize_debts(balances_list)
    
    return {
        "balances": balances_list,
        "debts": debts_list
    }

File: backend-spring/src/main/java/com/snapsplit/service/BalanceService.java
package com.snapsplit.service;

import com.snapsplit.dto.response.BalanceViewResponse;
import com.snapsplit.entity.*;
import com.snapsplit.repository.*;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.time.LocalDateTime;
import java.util.*;

/**
 * Balance computation service — equivalent to Python's balance_service.py
 *
 * Computes per-user balances (paid, share, settled, net) and
 * minimizes debts using a greedy algorithm.
 */
@Service
public class BalanceService {

    private final GroupMemberRepository groupMemberRepository;
    private final ExpenseRepository expenseRepository;
    private final ExpenseItemRepository expenseItemRepository;
    private final SplitRepository splitRepository;
    private final SettlementRepository settlementRepository;
    private final GroupBalanceRepository groupBalanceRepository;

    @PersistenceContext
    private EntityManager entityManager;

    public BalanceService(GroupMemberRepository groupMemberRepository,
            ExpenseRepository expenseRepository,
            ExpenseItemRepository expenseItemRepository,
            SplitRepository splitRepository,
            SettlementRepository settlementRepository,
            GroupBalanceRepository groupBalanceRepository) {
        this.groupMemberRepository = groupMemberRepository;
        this.expenseRepository = expenseRepository;
        this.expenseItemRepository = expenseItemRepository;
        this.splitRepository = splitRepository;
        this.settlementRepository = settlementRepository;
        this.groupBalanceRepository = groupBalanceRepository;
    }

    /**
     * Compute net balance for each user in a group.
     * Python equivalent: compute_group_balances()
     */
    public List<BalanceViewResponse.GroupBalanceDetail> computeGroupBalances(UUID groupId) {
        List<GroupMemberEntity> members = groupMemberRepository.findByGroupId(groupId);

        // Build balance map: userId → balance detail
        Map<UUID, BalanceViewResponse.GroupBalanceDetail> balances = new LinkedHashMap<>();
        for (GroupMemberEntity member : members) {
            UserEntity user = member.getUser();
            balances.put(member.getUserId(), BalanceViewResponse.GroupBalanceDetail.builder()
                    .userId(member.getUserId())
                    .userName(user.getName())
                    .totalPaid(BigDecimal.ZERO)
                    .totalShare(BigDecimal.ZERO)
                    .totalSettled(BigDecimal.ZERO)
                    .netBalance(BigDecimal.ZERO)
                    .updatedAt(LocalDateTime.now())
                    .build());
        }

        // Process expenses
        List<ExpenseEntity> expenses = expenseRepository.findByGroupIdOrderByCreatedAtDesc(groupId);
        for (ExpenseEntity expense : expenses) {
            if (expense.getTotalAmount() == null)
                continue;

            UUID payerId = expense.getCreator() != null ? expense.getCreator().getId() : null;
            BigDecimal expenseTotalShares = BigDecimal.ZERO;

            List<ExpenseItemEntity> items = expenseItemRepository.findByExpenseId(expense.getId());

            // Batch load all splits for performance (avoids N+1)
            if (!items.isEmpty()) {
                List<UUID> itemIds = items.stream().map(ExpenseItemEntity::getId).toList();
                List<SplitEntity> allSplits = splitRepository.findAllById(
                        itemIds.stream().map(id -> {
                            // Use findByExpenseItemId for each — simplified approach
                            return id;
                        }).toList());

                // Collect all splits for these items
                List<SplitEntity> splits = new ArrayList<>();
                for (ExpenseItemEntity item : items) {
                    splits.addAll(splitRepository.findByExpenseItemId(item.getId()));
                }

                for (SplitEntity split : splits) {
                    BalanceViewResponse.GroupBalanceDetail bal = balances.get(split.getUser().getId());
                    if (bal != null) {
                        bal.setTotalShare(bal.getTotalShare().add(split.getAmount()));
                        bal.setNetBalance(bal.getNetBalance().subtract(split.getAmount()));
                        expenseTotalShares = expenseTotalShares.add(split.getAmount());
                    }
                }
            }

            // Credit the payer with the exact amount that was split
            if (payerId != null && balances.containsKey(payerId)) {
                BalanceViewResponse.GroupBalanceDetail payerBal = balances.get(payerId);
                payerBal.setTotalPaid(payerBal.getTotalPaid().add(expenseTotalShares));
                payerBal.setNetBalance(payerBal.getNetBalance().add(expenseTotalShares));
            }
        }

        // Process settlements
        List<SettlementEntity> settlements = settlementRepository.findByGroupIdOrderByCreatedAtDesc(groupId);
        for (SettlementEntity settlement : settlements) {
            UUID payerById = settlement.getPayer().getId();
            UUID payeeId = settlement.getPayee().getId();

            if (balances.containsKey(payerById)) {
                BalanceViewResponse.GroupBalanceDetail b = balances.get(payerById);
                b.setNetBalance(b.getNetBalance().add(settlement.getAmount()));
                b.setTotalSettled(b.getTotalSettled().subtract(settlement.getAmount()));
            }
            if (balances.containsKey(payeeId)) {
                BalanceViewResponse.GroupBalanceDetail b = balances.get(payeeId);
                b.setNetBalance(b.getNetBalance().subtract(settlement.getAmount()));
                b.setTotalSettled(b.getTotalSettled().add(settlement.getAmount()));
            }
        }

        return new ArrayList<>(balances.values());
    }

    /**
     * Update the group_balances cache table.
     * Python equivalent: update_group_balances() with SELECT FOR UPDATE
     */
    @Transactional
    public void updateGroupBalances(UUID groupId) {
        // Row-level lock to prevent race conditions
        entityManager.createNativeQuery("SELECT * FROM group_balances WHERE group_id = :gid FOR UPDATE")
                .setParameter("gid", groupId)
                .getResultList();

        List<BalanceViewResponse.GroupBalanceDetail> computed = computeGroupBalances(groupId);

        for (BalanceViewResponse.GroupBalanceDetail detail : computed) {
            Optional<GroupBalanceEntity> existing = groupBalanceRepository
                    .findByGroupIdAndUserId(groupId, detail.getUserId());

            if (existing.isPresent()) {
                GroupBalanceEntity entity = existing.get();
                entity.setNetBalance(detail.getNetBalance());
                entity.setUpdatedAt(LocalDateTime.now());
                groupBalanceRepository.save(entity);
            } else {
                GroupBalanceEntity entity = GroupBalanceEntity.builder()
                        .groupId(groupId)
                        .userId(detail.getUserId())
                        .netBalance(detail.getNetBalance())
                        .updatedAt(LocalDateTime.now())
                        .build();
                groupBalanceRepository.save(entity);
            }
        }
    }

    /**
     * Minimize debts using a greedy algorithm.
     * Python equivalent: minimize_debts()
     */
    public List<BalanceViewResponse.DebtDetail> minimizeDebts(List<BalanceViewResponse.GroupBalanceDetail> balances) {
        List<Map.Entry<String, BigDecimal>> debtors = new ArrayList<>();
        List<Map.Entry<String, BigDecimal>> creditors = new ArrayList<>();

        for (BalanceViewResponse.GroupBalanceDetail b : balances) {
            int cmp = b.getNetBalance().compareTo(BigDecimal.ZERO);
            if (cmp < 0) {
                debtors.add(Map.entry(b.getUserName(), b.getNetBalance().negate()));
            } else if (cmp > 0) {
                creditors.add(Map.entry(b.getUserName(), b.getNetBalance()));
            }
        }

        // Sort descending by amount
        debtors.sort((a, b) -> b.getValue().compareTo(a.getValue()));
        creditors.sort((a, b) -> b.getValue().compareTo(a.getValue()));

        List<BalanceViewResponse.DebtDetail> debts = new ArrayList<>();
        // Use mutable wrappers
        BigDecimal[] debtorAmounts = debtors.stream().map(Map.Entry::getValue).toArray(BigDecimal[]::new);
        BigDecimal[] creditorAmounts = creditors.stream().map(Map.Entry::getValue).toArray(BigDecimal[]::new);

        int i = 0, j = 0;
        while (i < debtors.size() && j < creditors.size()) {
            BigDecimal amount = debtorAmounts[i].min(creditorAmounts[j]);

            if (amount.compareTo(BigDecimal.ZERO) > 0) {
                debts.add(BalanceViewResponse.DebtDetail.builder()
                        .fromUser(debtors.get(i).getKey())
                        .toUser(creditors.get(j).getKey())
                        .amount(amount.setScale(2, RoundingMode.HALF_UP))
                        .build());
            }

            debtorAmounts[i] = debtorAmounts[i].subtract(amount);
            creditorAmounts[j] = creditorAmounts[j].subtract(amount);

            if (debtorAmounts[i].compareTo(new BigDecimal("0.01")) < 0)
                i++;
            if (creditorAmounts[j].compareTo(new BigDecimal("0.01")) < 0)
                j++;
        }

        return debts;
    }

    /**
     * Calculate comprehensive group debts (balances + simplified debts).
     * Python equivalent: calculate_group_debts()
     */
    public BalanceViewResponse calculateGroupDebts(UUID groupId) {
        List<BalanceViewResponse.GroupBalanceDetail> balances = computeGroupBalances(groupId);
        List<BalanceViewResponse.DebtDetail> debts = minimizeDebts(balances);

        return BalanceViewResponse.builder()
                .balances(balances)
                .debts(debts)
                .build();
    }
}
